File: Raspberry_pi_approach/imu_read.cpp
// g++ -O2 -std=c++17 icm20948_spi.cpp -o icm
#include <linux/spi/spidev.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <unistd.h>

#include <cstdint>
#include <cstdio>
#include <cstring>
#include <chrono>

static constexpr uint8_t REG_BANK_SEL = 0x7F; // :contentReference[oaicite:20]{index=20}
static constexpr uint8_t WHO_AM_I     = 0x00; // :contentReference[oaicite:21]{index=21}
static constexpr uint8_t USER_CTRL    = 0x03; // :contentReference[oaicite:22]{index=22}
static constexpr uint8_t PWR_MGMT_1   = 0x06; // :contentReference[oaicite:23]{index=23}
static constexpr uint8_t PWR_MGMT_2   = 0x07; // :contentReference[oaicite:24]{index=24}
static constexpr uint8_t ACCEL_XOUT_H = 0x2D; // :contentReference[oaicite:25]{index=25}

static constexpr uint8_t ACCEL_SMPLRT_DIV_1 = 0x10; // bank2 :contentReference[oaicite:26]{index=26}
static constexpr uint8_t ACCEL_SMPLRT_DIV_2 = 0x11; // bank2 :contentReference[oaicite:27]{index=27}

static inline uint8_t bank_val(int bank) { return (uint8_t)((bank & 0x3) << 4); }

class Spi {
public:
    Spi(const char* dev, uint32_t hz) {
        fd_ = ::open(dev, O_RDWR);
        if (fd_ < 0) { perror("open"); std::exit(1); }

        uint8_t mode = SPI_MODE_0;
        uint8_t bits = 8;
        if (ioctl(fd_, SPI_IOC_WR_MODE, &mode) < 0) perror("SPI_IOC_WR_MODE");
        if (ioctl(fd_, SPI_IOC_WR_BITS_PER_WORD, &bits) < 0) perror("SPI_IOC_WR_BITS_PER_WORD");
        if (ioctl(fd_, SPI_IOC_WR_MAX_SPEED_HZ, &hz) < 0) perror("SPI_IOC_WR_MAX_SPEED_HZ");
        hz_ = hz;
    }

    ~Spi() { if (fd_ >= 0) ::close(fd_); }

    void xfer(const uint8_t* tx, uint8_t* rx, size_t n) {
        spi_ioc_transfer tr{};
        tr.tx_buf = (unsigned long)tx;
        tr.rx_buf = (unsigned long)rx;
        tr.len = n;
        tr.speed_hz = hz_;
        tr.bits_per_word = 8;
        tr.cs_change = 0;
        if (ioctl(fd_, SPI_IOC_MESSAGE(1), &tr) < 0) perror("SPI_IOC_MESSAGE");
    }

private:
    int fd_{-1};
    uint32_t hz_{0};
};

struct AccelRaw { int16_t ax, ay, az; };

class Icm20948 {
public:
    Icm20948(Spi& spi) : spi_(spi) {}

    void set_bank(int bank) { write_reg_raw(REG_BANK_SEL, bank_val(bank)); }

    uint8_t read_reg_raw(uint8_t reg) {
        uint8_t tx[2] = { (uint8_t)(0x80 | (reg & 0x7F)), 0x00 };
        uint8_t rx[2] = {};
        spi_.xfer(tx, rx, 2);
        return rx[1];
    }

    void write_reg_raw(uint8_t reg, uint8_t val) {
        uint8_t tx[2] = { (uint8_t)(reg & 0x7F), val };
        uint8_t rx[2] = {};
        spi_.xfer(tx, rx, 2);
    }

    void write_reg(int bank, uint8_t reg, uint8_t val) {
        set_bank(bank);
        write_reg_raw(reg, val);
    }

    void init_spi_only() {
        set_bank(0);

        // I2C_IF_DIS bit (bit4) in USER_CTRL :contentReference[oaicite:28]{index=28}
        uint8_t uc = read_reg_raw(USER_CTRL);
        uc |= (1u << 4);
        write_reg_raw(USER_CTRL, uc);

        write_reg_raw(PWR_MGMT_1, 0x01);
        write_reg_raw(PWR_MGMT_2, 0x00);
    }

    bool check_whoami() {
        set_bank(0);
        return read_reg_raw(WHO_AM_I) == 0xEA;
    }

    void set_accel_rate_div(uint16_t div12) {
        div12 &= 0x0FFF;
        write_reg(2, ACCEL_SMPLRT_DIV_1, (div12 >> 8) & 0x0F);
        write_reg(2, ACCEL_SMPLRT_DIV_2, div12 & 0xFF);
    }

    AccelRaw read_accel_raw() {
        set_bank(0);
        uint8_t tx[1 + 6];
        uint8_t rx[1 + 6];
        tx[0] = (uint8_t)(0x80 | (ACCEL_XOUT_H & 0x7F));
        std::memset(tx + 1, 0, 6);
        std::memset(rx, 0, sizeof(rx));
        spi_.xfer(tx, rx, sizeof(tx));

        auto be16 = [&](int i)->int16_t {
            return (int16_t)((rx[i] << 8) | rx[i+1]);
        };
        return { be16(1), be16(3), be16(5) };
    }

private:
    Spi& spi_;
};

int main() {
    Spi spi("/dev/spidev0.0", 7'000'000);
    Icm20948 imu(spi);

    imu.init_spi_only();
    if (!imu.check_whoami()) {
        std::fprintf(stderr, "WHO_AM_I mismatch\n");
        return 1;
    }
    std::puts("ICM-20948 OK");

    imu.set_accel_rate_div(0);

    const double target_hz = 1000.0;
    const auto period = std::chrono::duration<double>(1.0 / target_hz);
    auto next = std::chrono::steady_clock::now();

    for (int i = 0; i < 2000; ++i) {
        while (std::chrono::steady_clock::now() < next) {}
        next += std::chrono::duration_cast<std::chrono::steady_clock::duration>(period);

        auto a = imu.read_accel_raw();
        std::printf("%d,%d,%d\n", a.ax, a.ay, a.az);
    }
    return 0;
}

File: Raspberry_pi_approach/imu_read.py
import spidev 
import time
import csv

REG_BANK_SEK = 0x7F
WHO_AM_I = 0x00
USER_CTRL = 0x03
PWR_MGMT_1 = 0x06
PWR_MGMT_2 = 0x07

ACCEL_XOUT_H = 0x2D
ACCEL_SMPLRT_DIV_1 = 0x10
ACCEL_SMPLRT_DIV_2 = 0x11
ACCEL_CONFIG = 0x14

def bank_val(bank):
    return (bank &0x3) << 4

class ICM20948_SPI:
    def __init__(self, bus=0, device=0, hz = 7_000_000):
        self.spi = spidev.SpiDev()
        self.spi.open(bus, device)
        self.spi.mode = 0
        self.spi.max_speed_hz = hz
        self.current_bank = None
        self.spi.bits_per_word = 8
        
        try: 
            self.sou.lsbfirst = False 
        except Exception: 
            pass 
        self.set_bank(0)
    
    def close(self):
        self.spi.close()
    
    def xfer(self, output):
        return self.spi.xfer2(output)
    
    def set_bank(self, bank):
        self.wrrite_reg(REG_BANK_SEK, bank_val(bank))
    
    def read_reg_raw(self, reg):
        resp = self.xfer([0x80 | (reg & 0x7F), 0x00])
        resp = resp[1]
        return resp
    
    
    def write_reg_raw(self, reg: int, val: int):
        self.xfer([reg & 0x7F, val & 0xFF])

    def read_bytes(self, start_reg: int, n: int) -> bytes:
        # Burst read: send address then clock out n bytes
        resp = self.xfer([0x80 | (start_reg & 0x7F)] + [0x00] * n)
        return bytes(resp[1:])

    def read_reg(self, bank: int, reg: int) -> int:
        self.set_bank(bank)
        return self.read_reg_raw(reg)

    def write_reg(self, bank: int, reg: int, val: int):
        self.set_bank(bank)
        self.write_reg_raw(reg, val)

    def init_spi_only(self):
        # Disable I2C interface: set I2C_IF_DIS (bit 4) in USER_CTRL  :contentReference[oaicite:17]{index=17}
        self.set_bank(0)
        uc = self.read_reg_raw(USER_CTRL)
        uc |= (1 << 4)
        self.write_reg_raw(USER_CTRL, uc)

        # Wake device: clear sleep (you can refine clock source later)
        # PWR_MGMT_1 exists at 0x06 (bank0)  :contentReference[oaicite:18]{index=18}
        self.write_reg_raw(PWR_MGMT_1, 0x01)  # common: auto clock select, sleep=0
        self.write_reg_raw(PWR_MGMT_2, 0x00)  # enable accel+gyro

    def set_accel_rate_div(self, div12: int):
        # div12 is 12-bit across DIV_1 (high nibble) and DIV_2 (low byte)
        div12 &= 0x0FFF
        self.write_reg(2, ACCEL_SMPLRT_DIV_1, (div12 >> 8) & 0x0F)
        self.write_reg(2, ACCEL_SMPLRT_DIV_2, div12 & 0xFF)

    def read_accel_raw(self):
        self.set_bank(0)
        b = self.read_bytes(ACCEL_XOUT_H, 6)  # XH,XL,YH,YL,ZH,ZL  :contentReference[oaicite:19]{index=19}
        ax = int.from_bytes(b[0:2], byteorder="big", signed=True)
        ay = int.from_bytes(b[2:4], byteorder="big", signed=True)
        az = int.from_bytes(b[4:6], byteorder="big", signed=True)
        return ax, ay, az

def main():
    imu = ICM20948_SPI(bus=0, dev=0, hz=7_000_000)
    imu.init_spi_only()

    who = imu.read_reg(0, WHO_AM_I)
    if who != 0xEA:
        print(f"WHO_AM_I mismatch: 0x{who:02X} (expected 0xEA)")
        imu.close()
        return
    print("ICM-20948 detected (WHO_AM_I = 0xEA)")

    # For ~1kHz accel, you can usually leave ODR high and just sample at 1kHz in software.
    # If you want a divider: try 0 (max internal rate) then sample at 1kHz loop rate.
    imu.set_accel_rate_div(0)

    target_hz = 1000.0
    dt = 1.0 / target_hz

    with open("accel_1khz.csv", "w", newline="") as f:
        w = csv.writer(f)
        w.writerow(["t_ns", "ax_raw", "ay_raw", "az_raw"])

        next_t = time.perf_counter()
        for _ in range(10_000):  # ~10 seconds
            now = time.perf_counter()
            if now < next_t:
                # Busy-wait gives tighter timing than sleep() for 1kHz
                continue
            next_t += dt

            t_ns = time.time_ns()
            ax, ay, az = imu.read_accel_raw()
            w.writerow([t_ns, ax, ay, az])

    imu.close()
    print("Wrote accel_1khz.csv")

if __name__ == "__main__":
    main()
